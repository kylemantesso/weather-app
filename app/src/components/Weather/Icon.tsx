import React from "react";

export const Icon = ({
  icon,
  description,
}: {
  icon: string;
  description: string;
}) => (
  <img
      alt={description}
    src={`https://www.metaweather.com/static/img/weather/${icon}.svg`}
    height="60px"
  />
);
