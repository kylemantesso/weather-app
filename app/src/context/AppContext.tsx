import React, { useEffect, useState } from "react";
import {
  fetchWeather,
  transformWeatherResult,
} from "../services/MetaweatherService";
import { IWeather, IWeatherResult } from "../types/weather";

export type FormatType = "C" | "F";

export interface IAppState {
  location: number;
  weather: IWeather;
  weatherResult: IWeatherResult;
  setLocation: (location: number) => void;
  formatType: FormatType;
  setFormatType: (formatType: FormatType) => void;
}

export const AppContext = React.createContext<Partial<IAppState>>({});

export const useAppContext = () => React.useContext(AppContext) as IAppState;

export const AppContextProvider: React.FC = ({ children }) => {
  const [location, setLocation] = useState<IAppState["location"]>();
  const [weather, setWeather] = useState<IAppState["weather"]>();
  const [weatherResult, setWeatherResult] = useState<
    IAppState["weatherResult"]
  >();
  const [formatType, setFormatType] = useState<IAppState["formatType"]>("C");

  // on location change, fetch weather data
  useEffect(() => {
    if (location) {
      fetchWeather(location)
        .then((weatherResult) => {
          setWeatherResult(weatherResult);
          setWeather(transformWeatherResult(weatherResult, formatType));
        })
        .catch((error) => {
          console.log(error);
          // TODO: Show nice error message and prompt for retry
        });
    }
  }, [location]);

  // on format type change, transform weather result
  useEffect(() => {
    if (weatherResult) {
      setWeather(transformWeatherResult(weatherResult, formatType));
    }
  }, [formatType]);

  return (
    <AppContext.Provider
      value={{
        location,
        setLocation,
        weather,
        formatType,
        setFormatType,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};
