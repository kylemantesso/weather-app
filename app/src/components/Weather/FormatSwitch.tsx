import React from "react";

import { Box, Button } from "grommet";
import { FormatType, useAppContext } from "../../context/AppContext";

export const FormatSwitch = ({ formatType }: { formatType: FormatType }) => {
  const { setFormatType } = useAppContext();
  return (
    <Box direction="row" gap="small">
      <Button
        id="c"
        primary={formatType === "C"}
        label="°C"
        size="small"
        onClick={() => setFormatType("C")}
      />
      <Button
        id="f"
        primary={formatType === "F"}
        label="°F"
        size="small"
        onClick={() => setFormatType("F")}
      />
    </Box>
  );
};
