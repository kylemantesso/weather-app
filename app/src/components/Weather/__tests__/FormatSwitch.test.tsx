import React from "react";
import renderer from "react-test-renderer";
import { Icon } from "../Icon";
import { FormatSwitch } from "../FormatSwitch";

import * as AppContext from "../../../context/AppContext";

jest.mock("../../../context/AppContext");
const mockAppContext = AppContext as jest.Mocked<typeof AppContext>;
const mockSetFormatType = jest.fn();
mockAppContext.useAppContext.mockReturnValue({
  setFormatType: mockSetFormatType,
});

describe("FormatSwitch", () => {
  it("matches snapshot for C", () => {
    const component = renderer.create(<FormatSwitch formatType="C" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("matches snapshot for F", () => {
    const component = renderer.create(<FormatSwitch formatType="F" />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("switches format type", () => {
    const instance = renderer.create(<FormatSwitch formatType="F" />).root;
    renderer.act(() => {
      instance.findByProps({ id: "f" }).props.onClick();
    });
    expect(mockSetFormatType).toHaveBeenCalledWith("F");
  });
});
