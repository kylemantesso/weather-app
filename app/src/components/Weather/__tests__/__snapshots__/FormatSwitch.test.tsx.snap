// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormatSwitch matches snapshot for C 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 bIdaiR"
>
  <button
    className="StyledButton-sc-323bzc-0 kphDHx"
    id="c"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    째C
  </button>
  <div
    className="StyledBox__StyledBoxGap-sc-13pk1d4-1 ffjnxm"
  />
  <button
    className="StyledButton-sc-323bzc-0 kFNfEv"
    id="f"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    째F
  </button>
</div>
`;

exports[`FormatSwitch matches snapshot for F 1`] = `
<div
  className="StyledBox-sc-13pk1d4-0 bIdaiR"
>
  <button
    className="StyledButton-sc-323bzc-0 kFNfEv"
    id="c"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    째C
  </button>
  <div
    className="StyledBox__StyledBoxGap-sc-13pk1d4-1 ffjnxm"
  />
  <button
    className="StyledButton-sc-323bzc-0 kphDHx"
    id="f"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onMouseOut={[Function]}
    onMouseOver={[Function]}
    type="button"
  >
    째F
  </button>
</div>
`;
